Fix sizeof of structs with a trailing unbound array

Expected 
```c
struct X_char {
  char i;
  char data[];
};

struct X_short {
  char i;
  short data[];
};


static_assert(sizeof(struct X_char) == 1);
static_assert(alignof(struct X_char) == 1);

static_assert(sizeof(struct X_short) == 2);
static_assert(alignof(struct X_short) == alignof(short));
```


```bash
cxx -xc x.c  -fcheck
x.c:2:8: error: size of incomplete type 'char data[]'
struct X_char {
       ^
x.c:7:8: error: size of incomplete type 'short data[]'
struct X_short {
       ^
x.c:13:1: error: static assert failed
static_assert(sizeof(struct X_char) == 1);
^
x.c:16:1: error: static assert failed
static_assert(sizeof(struct X_short) == 2);
^
x.c:17:1: error: static assert failed
static_assert(alignof(struct X_short) == alignof(short));
^
```

--- a/src/parser/cxx/symbols.cc
+++ b/src/parser/cxx/symbols.cc
@@ -290,10 +290,20 @@ auto ClassSymbol::buildClassLayout(Control* control)
     alignment = std::max(alignment, baseClassSymbol->alignment());
   }
 
+  FieldSymbol* lastField = nullptr;
+
   for (auto member : scope()->symbols()) {
     auto field = symbol_cast<FieldSymbol>(member);
     if (!field) continue;
     if (field->isStatic()) continue;
+
+    if (lastField && control->is_unbounded_array(lastField->type())) {
+      // If the last field is an unbounded array, we cannot compute the offset
+      // of the next field, so we report an error.
+      return std::unexpected(
+          std::format("size of incomplete type '{}'",
+                      to_string(lastField->type(), lastField->name())));
+    }
 
     if (!field->alignment()) {
       return std::unexpected(
@@ -301,7 +311,13 @@ auto ClassSymbol::buildClassLayout(Control* control)
                       to_string(field->type(), field->name())));
     }
 
-    auto size = memoryLayout->sizeOf(field->type());
+    std::optional<std::size_t> size;
+
+    if (control->is_unbounded_array(field->type())) {
+      size = 0;
+    } else {
+      size = memoryLayout->sizeOf(field->type());
+    }
 
     if (!size.has_value()) {
       return std::unexpected(
@@ -318,6 +334,8 @@ auto ClassSymbol::buildClassLayout(Control* control)
     }
 
     alignment = std::max(alignment, field->alignment());
+
+    lastField = field;
   }
 
   offset = align_to(offset, alignment);

--- a/tests/unit_tests/sema/sizeof_01.cc
+++ b/tests/unit_tests/sema/sizeof_01.cc
@@ -28,3 +28,10 @@ static_assert(sizeof(elements) == 5 * sizeof(int));
 
 double d[]{1.0, 2.0, 3.0};
 static_assert(sizeof(d) == 3 * sizeof(double));
+
+struct X {
+  char i;
+  int dat[];
+};
+static_assert(sizeof(X) == 4);
+static_assert(alignof(X) == 4);

